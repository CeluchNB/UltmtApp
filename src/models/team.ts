import { DisplayUser } from '../types/user'
import { Realm } from '@realm/react'
import { LocalUser, Team } from '../types/team'

export const DisplayTeamSchema: Realm.ObjectSchema = {
    name: 'DisplayTeam',
    embedded: true,
    properties: {
        _id: 'string',
        place: 'string',
        name: 'string',
        teamname: 'string',
        seasonStart: 'string',
        seasonEnd: 'string',
    },
}

export const GuestTeamSchema: Realm.ObjectSchema = {
    name: 'GuestTeam',
    embedded: true,
    properties: {
        _id: 'string?',
        place: 'string?',
        name: 'string',
        teamname: 'string?',
        seasonStart: 'string?',
        seasonEnd: 'string?',
    },
}

export class TeamSchema {
    static schema: Realm.ObjectSchema = {
        name: 'Team',
        primaryKey: '_id',
        properties: {
            _id: 'string',
            place: 'string',
            name: 'string',
            teamname: 'string',
            managers: { type: 'list', objectType: 'DisplayUser' },
            players: { type: 'list', objectType: 'DisplayUser' },
            seasonStart: 'string',
            seasonEnd: 'string',
            seasonNumber: 'int',
            continuationId: 'string',
            rosterOpen: 'bool',
        },
    }

    _id: string
    place: string
    name: string
    teamname: string
    managers: DisplayUser[]
    players: LocalUser[]
    seasonStart: string
    seasonEnd: string
    seasonNumber: number
    continuationId: string
    rosterOpen: boolean

    constructor(team: Team) {
        this._id = team._id
        this.place = team.place
        this.name = team.name
        this.teamname = team.teamname
        this.managers = team.managers
        this.players = team.players
        this.seasonStart = team.seasonStart
        this.seasonEnd = team.seasonEnd
        this.seasonNumber = team.seasonNumber
        this.continuationId = team.continuationId
        this.rosterOpen = team.rosterOpen
    }
}
