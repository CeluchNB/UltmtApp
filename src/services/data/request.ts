import * as Constants from '../../utils/constants'
import { DetailedRequest } from '../../types/request'
import { throwApiError } from '../../utils/service-utils'
import {
    deleteTeamRequest as networkDeleteTeamRequest,
    deleteUserRequest as networkDeleteUserRequest,
    getRequest as networkGetRequest,
    getRequestsByTeam as networkGetRequestsByTeam,
    getRequestsByUser as networkGetRequestsByUser,
    requestTeam as networkRequestTeam,
    requestUser as networkRequestUser,
    respondToPlayerRequest as networkRespondToPlayerRequest,
    respondToTeamRequest as networkRespondToTeamRequest,
} from '../network/request'

/**
 * Method for user to request to join a team.
 * @param token jwt of requesting user
 * @param teamId id of requested team
 * @returns a detailed request object
 * @throws error if backend returns an error
 */
export const requestTeam = async (
    token: string,
    teamId: string,
): Promise<DetailedRequest> => {
    try {
        const response = await networkRequestTeam(token, teamId)
        const { request } = response.data
        return request
    } catch (error) {
        return throwApiError(error, Constants.CREATE_REQUEST_ERROR)
    }
}

/**
 * Method for team to request a user join a team.
 * @param token jwt of team manager
 * @param userId user to request join team
 * @param teamId team requesting the player to join
 * @returns a detailed request object
 * @throws error if backend returns an error
 */
export const requestUser = async (
    token: string,
    userId: string,
    teamId: string,
): Promise<DetailedRequest> => {
    try {
        const response = await networkRequestUser(token, userId, teamId)
        const { request } = response.data
        return request
    } catch (error) {
        return throwApiError(error, Constants.CREATE_REQUEST_ERROR)
    }
}

/**
 * Method to get request details. User must be authorized to view request.
 * User is authorized if 1) he is the user on the request or 2) he is a manager
 * of the team on the request
 * @param token jwt of current user
 * @param requestId id of request
 * @returns a detailed request object
 * @throws error if backend returns an error
 */
export const getRequest = async (
    token: string,
    requestId: string,
): Promise<DetailedRequest> => {
    try {
        const response = await networkGetRequest(token, requestId)
        const { request } = response.data
        return request
    } catch (error) {
        return throwApiError(error, Constants.GET_REQUEST_ERROR)
    }
}

/**
 * Method for team to respond to a player's request
 * @param token jwt of team manager
 * @param requestId id of request
 * @param accept boolean for accept/deny
 * @returns a detailed request object
 * @throws error if backend returns an error
 */
export const respondToPlayerRequest = async (
    token: string,
    requestId: string,
    accept: boolean,
): Promise<DetailedRequest> => {
    try {
        const response = await networkRespondToPlayerRequest(
            token,
            requestId,
            accept,
        )
        const { request } = response.data
        return request
    } catch (error) {
        return throwApiError(error, Constants.REQUEST_RESPONSE_ERROR)
    }
}

/**
 * Method for a team to delete a pending request it sent.
 * @param token jwt of team manager
 * @param requestId id of request to delete
 * @returns a detailed request object
 * @throws error if backend returns an error
 */
export const deleteTeamRequest = async (
    token: string,
    requestId: string,
): Promise<DetailedRequest> => {
    try {
        const response = await networkDeleteTeamRequest(token, requestId)
        const { request } = response.data
        return request
    } catch (error) {
        return throwApiError(error, Constants.REQUEST_RESPONSE_ERROR)
    }
}

/**
 * Method for user to respond to a request from a team
 * @param token jwt of user
 * @param requestId id of request to respond to
 * @param accept boolean for accept/deny
 * @returns a detailed request object
 * @throws error if backend returns an error
 */
export const respondToTeamRequest = async (
    token: string,
    requestId: string,
    accept: boolean,
): Promise<DetailedRequest> => {
    try {
        const response = await networkRespondToTeamRequest(
            token,
            requestId,
            accept,
        )
        const { request } = response.data
        return request
    } catch (error) {
        return throwApiError(error, Constants.REQUEST_RESPONSE_ERROR)
    }
}

/**
 * Method for a user to delete a pending request she has sent.
 * @param token jwt of user
 * @param requestId id of request to delete
 * @returns a detailed request object
 * @throws error if backend returns an error
 */
export const deleteUserRequest = async (
    token: string,
    requestId: string,
): Promise<DetailedRequest> => {
    try {
        const response = await networkDeleteUserRequest(token, requestId)
        const { request } = response.data
        return request
    } catch (error) {
        return throwApiError(error, Constants.REQUEST_RESPONSE_ERROR)
    }
}

/**
 * Get all requests belonging to a team
 * @param token token of team manager
 * @param teamId id of team to get requests from
 * @returns list of detailed requests
 */
export const getRequestsByTeam = async (
    token: string,
    teamId: string,
): Promise<DetailedRequest[]> => {
    try {
        const response = await networkGetRequestsByTeam(token, teamId)
        const { requests } = response.data
        return requests
    } catch (error) {
        return throwApiError(error, Constants.REQUEST_RESPONSE_ERROR)
    }
}

/**
 * Get all requests beloinging to a user
 * @param token token of user to get requests for
 * @returns list of detailed requests
 */
export const getRequestsByUser = async (
    token: string,
): Promise<DetailedRequest[]> => {
    try {
        const response = await networkGetRequestsByUser(token)
        const { requests } = response.data
        return requests
    } catch (error) {
        return throwApiError(error, Constants.REQUEST_RESPONSE_ERROR)
    }
}
