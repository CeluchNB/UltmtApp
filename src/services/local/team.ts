import { Realm } from '@realm/react'
import { Team } from '../../types/team'
import { TeamSchema } from '../../models'
import { getRealm } from '../../models/realm'

export const saveTeams = async (teams: Team[]) => {
    const realm = await getRealm()
    realm.write(() => {
        for (const team of teams) {
            realm.create('Team', team, Realm.UpdateMode.Modified)
        }
    })
}

export const getManagingTeams = async (id: string): Promise<Team[]> => {
    const realm = await getRealm()
    const teams = await realm.objects<TeamSchema>('Team')
    return teams
        .filter(t => t.managers.map(m => m._id).includes(id))
        .map(t => parseTeam(t))
}

const parseTeam = (schema: TeamSchema): Team => {
    return JSON.parse(
        JSON.stringify({
            _id: schema._id,
            place: schema.place,
            name: schema.name,
            teamname: schema.teamname,
            managers: schema.managers,
            players: schema.players,
            seasonStart: schema.seasonStart,
            seasonEnd: schema.seasonEnd,
            seasonNumber: schema.seasonNumber,
            continuationId: schema.continuationId,
            rosterOpen: schema.rosterOpen,
            requests: [],
            games: [],
        }),
    )
}
