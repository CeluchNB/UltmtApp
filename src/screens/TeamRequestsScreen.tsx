import * as React from 'react'
import * as RequestData from '../services/data/request'
import MapSection from '../components/molecules/MapSection'
import PrimaryButton from '../components/atoms/PrimaryButton'
import { Props } from '../types/navigation'
import ScreenTitle from '../components/atoms/ScreenTitle'
import UserListItem from '../components/atoms/UserListItem'
import { size } from '../theme/fonts'
import { useColors } from '../hooks'
import { DetailedRequest, RequestType } from '../types/request'
import {
    RefreshControl,
    SafeAreaView,
    ScrollView,
    StyleSheet,
    Text,
    View,
} from 'react-native'
import {
    getManagedTeam,
    selectOpenLoading,
    selectTeam,
    toggleRosterStatus,
} from '../store/reducers/features/team/managedTeamReducer'
import { useDispatch, useSelector } from 'react-redux'

const TeamRequestsScreen: React.FC<Props> = ({ navigation }) => {
    const dispatch = useDispatch()
    const { colors } = useColors()

    const team = useSelector(selectTeam)
    const openLoading = useSelector(selectOpenLoading)

    const isMounted = React.useRef(false)
    const [refresh, setRefresh] = React.useState(false)
    const [error, setError] = React.useState<string>('')
    const [requestsLoading, setRequestsLoading] = React.useState(false)
    const [requests, setRequests] = React.useState([] as DetailedRequest[])
    const [deleteRequestError, setDeleteRequestError] = React.useState('')
    const [deleteRequestId, setDeleteRequestId] = React.useState('')
    const [respondRequestError, setRespondRequestError] = React.useState('')
    const [respondRequestId, setResponseRequestId] = React.useState('')

    const initializeScreen = React.useCallback(async () => {
        try {
            setError('')
            if (!team) {
                setError('No team data. Please refresh or try again later.')
                return
            }
            setRequestsLoading(true)
            const reqs = await RequestData.getRequestsByTeam(team._id)
            if (isMounted.current) {
                setRequests(reqs)
            }
        } catch (e: any) {
            setError(e.message)
        } finally {
            if (isMounted.current) {
                setRequestsLoading(false)
            }
        }
    }, [team])

    React.useEffect(() => {
        isMounted.current = true
        const unsubscribe = navigation.addListener('focus', () => {
            initializeScreen()
        })
        return () => {
            unsubscribe()
            isMounted.current = false
        }
    })

    const respondToRequest = async (requestId: string, accept: boolean) => {
        try {
            setRespondRequestError('')
            setResponseRequestId(requestId)
            await RequestData.respondToPlayerRequest(requestId, accept)
            setRequests(requests.filter(r => r._id !== requestId))

            if (accept) {
                dispatch(getManagedTeam({ id: team?._id || '' }))
            }
        } catch (e: any) {
            console.log('in screen', e)
            setRespondRequestError(
                e.message ?? 'Unable to respond to this request',
            )
        }
    }

    const deleteRequest = async (requestId: string) => {
        try {
            setDeleteRequestError('')
            setDeleteRequestId(requestId)
            const request = await RequestData.deleteTeamRequest(requestId)
            // ONLY FILTERING LOCALLY, SHOULD I RE-CALL 'getTeam'?
            setRequests(requests.filter(r => r._id !== request._id))
        } catch (e: any) {
            setDeleteRequestError(e.message ?? 'Unable to delete this request.')
        }
    }

    const styles = StyleSheet.create({
        screen: {
            height: '100%',
            backgroundColor: colors.primary,
            flex: 1,
        },
        title: {
            textAlign: 'center',
        },
        error: {
            color: colors.error,
            fontSize: size.fontLarge,
        },
        toggleRosterButton: {
            alignSelf: 'center',
        },
        container: {
            alignSelf: 'center',
            width: '75%',
        },
    })

    return (
        <SafeAreaView style={styles.screen}>
            <ScrollView
                refreshControl={
                    <RefreshControl
                        refreshing={refresh}
                        onRefresh={async () => {
                            setRefresh(true)
                            await initializeScreen()
                            setRefresh(false)
                        }}
                    />
                }
                testID="mtd-flat-list">
                <ScreenTitle
                    title={`${team?.name} Requests`}
                    style={styles.title}
                />
                {error.length > 0 && <Text style={styles.error}>{error}</Text>}
                <PrimaryButton
                    style={styles.toggleRosterButton}
                    text={`${team?.rosterOpen ? 'Close' : 'Open'} Roster`}
                    loading={openLoading}
                    onPress={async () => {
                        dispatch(
                            toggleRosterStatus({
                                id: team?._id || '',
                                open: !team?.rosterOpen,
                            }),
                        )
                    }}
                />
                <View style={styles.container}>
                    <MapSection
                        title="Request From Players"
                        listData={requests.filter(
                            item => item.requestSource !== 'team',
                        )}
                        renderItem={item => {
                            return (
                                <UserListItem
                                    key={item._id}
                                    user={
                                        item.userDetails || {
                                            firstName: 'User no',
                                            lastName: 'longer exists',
                                            username: 'deleteme',
                                        }
                                    }
                                    showDelete={true}
                                    showAccept={true}
                                    onDelete={() =>
                                        respondToRequest(item._id, false)
                                    }
                                    onAccept={() =>
                                        respondToRequest(item._id, true)
                                    }
                                    error={
                                        respondRequestError.length > 0 &&
                                        item._id === respondRequestId
                                            ? respondRequestError
                                            : undefined
                                    }
                                />
                            )
                        }}
                        loading={requestsLoading}
                        showButton={false}
                        showCreateButton={false}
                        buttonText=""
                        onButtonPress={() => {}}
                        error={
                            requests.filter(
                                item => item.requestSource !== 'team',
                            ).length <= 0
                                ? 'No open requests from players'
                                : undefined
                        }
                    />
                    <MapSection
                        title="Requests To Players"
                        listData={requests.filter(
                            item => item.requestSource !== 'player',
                        )}
                        renderItem={item => {
                            return (
                                <UserListItem
                                    key={item._id}
                                    user={
                                        item.userDetails || {
                                            firstName: 'User no',
                                            lastName: 'longer exists',
                                            username: 'deleteme',
                                        }
                                    }
                                    showDelete={true}
                                    showAccept={false}
                                    requestStatus={item.status}
                                    onDelete={() => deleteRequest(item._id)}
                                    error={
                                        item._id === deleteRequestId &&
                                        deleteRequestError.length > 0
                                            ? deleteRequestError
                                            : undefined
                                    }
                                />
                            )
                        }}
                        loading={requestsLoading}
                        showButton={false}
                        showCreateButton={true}
                        onCreatePress={() => {
                            navigation.navigate('RequestUser', {
                                type: RequestType.PLAYER,
                            })
                        }}
                        error={
                            requests.filter(
                                item => item.requestSource !== 'player',
                            ).length <= 0
                                ? 'No open requests to players'
                                : undefined
                        }
                    />
                </View>
            </ScrollView>
        </SafeAreaView>
    )
}

export default TeamRequestsScreen
