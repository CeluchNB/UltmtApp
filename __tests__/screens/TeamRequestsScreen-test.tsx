import * as ManagedTeamReducer from '../../src/store/reducers/features/team/managedTeamReducer'
import * as RequestData from '../../src/services/data/request'
import * as TeamData from '../../src/services/data/team'
import { DetailedRequest } from '../../src/types/request'
import { NavigationContainer } from '@react-navigation/native'
import { Provider } from 'react-redux'
import React from 'react'
import { RequestType } from '../../src/types/request'
import { Team } from '../../src/types/team'
import { TeamRequestProps } from '../../src/types/navigation'
import TeamRequestsScreen from '../../src/screens/TeamRequestsScreen'
import store from '../../src/store/store'
import { waitUntilRefreshComplete } from '../../fixtures/utils'
import { act, fireEvent, render, waitFor } from '@testing-library/react-native'

jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper')

const navigate = jest.fn()
const addListener = jest.fn().mockReturnValue(() => {})
const setOptions = jest.fn()

const props: TeamRequestProps = {
    navigation: {
        navigate,
        addListener,
        setOptions,
    } as any,
    route: {} as any,
}

let getManagedTeamResponse: Team
let requestObject: DetailedRequest

beforeEach(() => {
    jest.clearAllMocks()
    getManagedTeamResponse = {
        _id: 'id1',
        place: 'place',
        name: 'name',
        teamname: 'placename',
        players: [
            {
                _id: 'playerid1',
                firstName: 'first1',
                lastName: 'last1',
                username: 'first1last1',
            },
        ],
        requests: ['request1', 'request2'],
        managers: [
            {
                _id: 'playerid4',
                firstName: 'first4',
                lastName: 'last4',
                username: 'first4last4',
            },
        ],
        seasonNumber: 1,
        seasonStart: '2022',
        seasonEnd: '2022',
        continuationId: 'id1234',
        rosterOpen: false,
        games: [],
    }
    requestObject = {
        _id: 'request1',
        team: 'id1',
        user: 'playerid1',
        requestSource: 'team',
        teamDetails: {
            _id: 'id1',
            place: 'place',
            name: 'name',
            teamname: 'placename',
            seasonStart: '2022',
            seasonEnd: '2022',
        },
        userDetails: {
            _id: 'playerid1',
            firstName: 'first1',
            lastName: 'last1',
            username: 'first1last1',
        },
        status: 'pending',
    }
    jest.spyOn(RequestData, 'getRequestsByTeam').mockReturnValue(
        Promise.resolve([
            {
                ...requestObject,
                _id: 'request1',
                requestSource: 'team',
                userDetails: {
                    _id: 'playerid2',
                    firstName: 'first2',
                    lastName: 'last2',
                    username: 'first2last2',
                },
            },
            {
                ...requestObject,
                _id: 'request2',
                requestSource: 'player',
                userDetails: {
                    _id: 'playerid3',
                    firstName: 'first3',
                    lastName: 'last3',
                    username: 'first3last3',
                },
            },
        ]),
    )
    jest.spyOn(TeamData, 'getManagedTeam').mockReturnValue(
        Promise.resolve(getManagedTeamResponse),
    )
    store.dispatch(ManagedTeamReducer.setTeam(getManagedTeamResponse))
})

it('should match snapshot', async () => {
    const snapshot = render(
        <Provider store={store}>
            <NavigationContainer>
                <TeamRequestsScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    await waitUntilRefreshComplete(snapshot.getByTestId('mtd-flat-list'))

    expect(snapshot.toJSON()).toMatchSnapshot()
})

it('should handle get request error', async () => {
    const spy = jest
        .spyOn(RequestData, 'getRequestsByTeam')
        .mockReturnValueOnce(Promise.reject({ message: 'test error' }))

    const { queryByText, getByTestId } = render(
        <Provider store={store}>
            <NavigationContainer>
                <TeamRequestsScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    await waitUntilRefreshComplete(getByTestId('mtd-flat-list'))

    expect(spy).toHaveBeenCalled()
    expect(queryByText('test error')).not.toBeNull()
})

it('should toggle roster status', async () => {
    const dataFn = jest
        .fn()
        .mockImplementationOnce(async (_token, id, open) => {
            return { rosterOpen: open } as Team
        })
    const spy = jest.spyOn(ManagedTeamReducer, 'toggleRosterStatus')

    jest.spyOn(TeamData, 'toggleRosterStatus').mockImplementationOnce(dataFn)

    const { getByText, queryByText, getByTestId } = render(
        <Provider store={store}>
            <NavigationContainer>
                <TeamRequestsScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    await waitUntilRefreshComplete(getByTestId('mtd-flat-list'))

    const button = getByText('Open Roster')
    fireEvent.press(button)

    waitFor(() => queryByText('Close Roster'))
    expect(spy).toHaveBeenCalledTimes(1)
    expect(dataFn).toHaveBeenCalledTimes(1)
})

it('should respond to request correctly', async () => {
    const responseSpy = jest
        .spyOn(RequestData, 'respondToPlayerRequest')
        .mockReturnValueOnce(
            Promise.resolve({
                ...requestObject,
                requestSource: 'player',
                userDetails: {
                    _id: 'playerid3',
                    firstName: 'first3',
                    lastName: 'last3',
                    username: 'first3last3',
                },
            }),
        )

    const { getByTestId, getAllByTestId, queryByText } = render(
        <Provider store={store}>
            <NavigationContainer>
                <TeamRequestsScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    await waitUntilRefreshComplete(getByTestId('mtd-flat-list'))

    jest.spyOn(RequestData, 'getRequestsByTeam').mockReturnValue(
        Promise.resolve([
            {
                ...requestObject,
                _id: 'request1',
                requestSource: 'team',
                userDetails: {
                    _id: 'playerid2',
                    firstName: 'first2',
                    lastName: 'last2',
                    username: 'first2last2',
                },
            },
        ]),
    )

    const acceptButtons = getAllByTestId('accept-button')
    expect(queryByText('@first3last3')).not.toBeNull()

    fireEvent.press(acceptButtons[0])
    // Not optimal solution, see
    // ManageTeamDetailsScreen-test.tsx for further details
    await act(async () => {})

    expect(queryByText('@first3last3')).toBeNull()
    expect(responseSpy).toHaveBeenCalledWith('request2', true)
})

it('should handle respond to request error', async () => {
    jest.spyOn(RequestData, 'respondToPlayerRequest').mockReturnValueOnce(
        Promise.reject({
            message: 'respond test error',
        }),
    )

    const { queryByText, getByTestId, getAllByTestId } = render(
        <Provider store={store}>
            <NavigationContainer>
                <TeamRequestsScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    await waitUntilRefreshComplete(getByTestId('mtd-flat-list'))

    const acceptButtons = getAllByTestId('accept-button')
    fireEvent.press(acceptButtons[0])
    // Not optimal solution, see
    // ManageTeamDetailsScreen-test.tsx for further details
    await act(async () => {})

    const errorText = queryByText('respond test error')
    expect(errorText).not.toBeNull()
})

it('should handle deny request', async () => {
    const responseSpy = jest
        .spyOn(RequestData, 'respondToPlayerRequest')
        .mockReturnValueOnce(
            Promise.resolve({
                ...requestObject,
                requestSource: 'player',
                userDetails: {
                    _id: 'playerid3',
                    firstName: 'first3',
                    lastName: 'last3',
                    username: 'first3last3',
                },
            }),
        )

    const { queryByText, getByTestId, getAllByTestId } = render(
        <Provider store={store}>
            <NavigationContainer>
                <TeamRequestsScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    await waitUntilRefreshComplete(getByTestId('mtd-flat-list'))

    jest.spyOn(RequestData, 'getRequestsByTeam').mockReturnValue(
        Promise.resolve([
            {
                ...requestObject,
                _id: 'request1',
                requestSource: 'team',
                userDetails: {
                    _id: 'playerid2',
                    firstName: 'first2',
                    lastName: 'last2',
                    username: 'first2last2',
                },
            },
        ]),
    )

    const deleteButtons = getAllByTestId('delete-button')
    fireEvent.press(deleteButtons[0])
    // Not optimal solution, see
    // ManageTeamDetailsScreen-test.tsx for further details
    await act(async () => {})

    expect(queryByText('@first3last3')).toBeNull()

    expect(responseSpy).toHaveBeenCalledWith('request2', false)
})

it('should handle delete request correctly', async () => {
    const deleteSpy = jest
        .spyOn(RequestData, 'deleteTeamRequest')
        .mockImplementationOnce(async () => {
            return requestObject
        })

    const { queryByText, getByTestId, getAllByTestId } = render(
        <Provider store={store}>
            <NavigationContainer>
                <TeamRequestsScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    await waitUntilRefreshComplete(getByTestId('mtd-flat-list'))

    jest.spyOn(RequestData, 'getRequestsByTeam').mockReturnValue(
        Promise.resolve([
            {
                ...requestObject,
                _id: 'request2',
                requestSource: 'player',
                userDetails: {
                    _id: 'playerid3',
                    firstName: 'first3',
                    lastName: 'last3',
                    username: 'first3last3',
                },
            },
        ]),
    )

    const username = '@first2last2'
    expect(queryByText(username)).not.toBeNull()
    const deleteButtons = getAllByTestId('delete-button')
    // Not optimal solution, see
    // ManageTeamDetailsScreen-test.tsx for further details
    fireEvent.press(deleteButtons[1])
    await act(async () => {})

    expect(deleteSpy).toHaveBeenCalled()
    expect(queryByText(username)).toBeNull()
})

it('should handle delete request error correctly', async () => {
    jest.spyOn(RequestData, 'deleteTeamRequest').mockReturnValueOnce(
        Promise.reject({
            message: 'delete error message',
        }),
    )

    const { queryByText, getByTestId, getAllByTestId } = render(
        <Provider store={store}>
            <NavigationContainer>
                <TeamRequestsScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    await waitUntilRefreshComplete(getByTestId('mtd-flat-list'))

    const username = '@first2last2'
    expect(queryByText(username)).not.toBeNull()
    const deleteButtons = getAllByTestId('delete-button')
    // Not optimal solution, see
    // ManageTeamDetailsScreen-test.tsx for further details
    fireEvent.press(deleteButtons[1])
    await act(async () => {})

    expect(queryByText(username)).not.toBeNull()
    expect(queryByText('delete error message')).not.toBeNull()
})

it('should handle navigate to request player', async () => {
    const { getByTestId } = render(
        <Provider store={store}>
            <NavigationContainer>
                <TeamRequestsScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    await waitUntilRefreshComplete(getByTestId('mtd-flat-list'))

    const button = getByTestId('create-button')
    fireEvent.press(button)

    expect(navigate).toHaveBeenCalledWith('RequestUser', {
        type: RequestType.PLAYER,
    })
})
