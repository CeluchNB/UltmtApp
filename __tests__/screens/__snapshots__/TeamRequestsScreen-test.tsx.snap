// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#ffffff",
      "flex": 1,
      "height": "100%",
    }
  }
>
  <RCTScrollView
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    testID="mtd-flat-list"
  >
    <RCTRefreshControl />
    <View>
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#3183ff",
              "fontSize": 61.64383561643835,
              "fontWeight": "bold",
            },
            Object {
              "textAlign": "center",
            },
          ]
        }
      >
        name Requests
      </Text>
      <View
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "#3183ff",
            "borderColor": "transparent",
            "borderRadius": 4,
            "borderStyle": "solid",
            "borderWidth": 0,
            "elevation": 2,
            "minWidth": "auto",
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 0.75,
              "width": 0,
            },
            "shadowOpacity": 0.24,
            "shadowRadius": 1.5,
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 4,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "letterSpacing": 1,
                      "marginHorizontal": 16,
                      "marginVertical": 9,
                      "textAlign": "center",
                    },
                    Object {
                      "marginHorizontal": 8,
                    },
                    Object {
                      "textTransform": "uppercase",
                    },
                    Object {
                      "color": "#ffffff",
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    Object {
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    undefined,
                  ],
                ]
              }
            >
              Open Roster
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "alignSelf": "center",
            "width": "75%",
          }
        }
      >
        <View
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <View
            style={
              Object {
                "display": "flex",
                "flex": 1,
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#3183ff",
                  "flex": 1,
                  "fontSize": 61.64383561643835,
                  "fontWeight": "bold",
                  "marginBottom": 10,
                }
              }
            >
              Request From Players
            </Text>
          </View>
          <Text
            style={
              Object {
                "fontSize": 41.0958904109589,
                "marginBottom": 50,
              }
            }
          >
            No open requests from players
          </Text>
        </View>
        <View
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <View
            style={
              Object {
                "display": "flex",
                "flex": 1,
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#3183ff",
                  "flex": 1,
                  "fontSize": 61.64383561643835,
                  "fontWeight": "bold",
                  "marginBottom": 10,
                }
              }
            >
              Requests To Players
            </Text>
            <View
              accessibilityRole="button"
              accessibilityState={
                Object {
                  "disabled": false,
                }
              }
              accessible={true}
              focusable={true}
              hitSlop={
                Object {
                  "bottom": 6,
                  "left": 6,
                  "right": 6,
                  "top": 6,
                }
              }
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  Object {
                    "overflow": "hidden",
                  },
                  Array [
                    Object {
                      "alignItems": "center",
                      "justifyContent": "center",
                      "margin": 6,
                      "overflow": "hidden",
                    },
                    Object {
                      "borderRadius": 18,
                      "height": 36,
                      "width": 36,
                    },
                    undefined,
                    Object {
                      "backgroundColor": "#3183ff",
                      "borderRadius": 5,
                      "display": "flex",
                      "flexWrap": "nowrap",
                    },
                  ],
                ]
              }
              testID="create-button"
            >
              <View>
                <Text
                  accessibilityElementsHidden={true}
                  allowFontScaling={false}
                  importantForAccessibility="no-hide-descendants"
                  pointerEvents="none"
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "color": "#ffffff",
                        "fontSize": 24,
                      },
                      Array [
                        Object {
                          "lineHeight": 24,
                          "transform": Array [
                            Object {
                              "scaleX": 1,
                            },
                          ],
                        },
                        Object {
                          "backgroundColor": "transparent",
                        },
                      ],
                      Object {
                        "fontFamily": "Material Design Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                >
                  Û∞êï
                </Text>
              </View>
            </View>
          </View>
          <Text
            style={
              Object {
                "fontSize": 41.0958904109589,
                "marginBottom": 50,
              }
            }
          >
            No open requests to players
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
